mint-client/src/clients/gateway.rs

```diff
@@ -216,6 +220,79 @@ impl GatewayClient {
         Ok(OutPoint { txid, out_idx: 0 })
     }
 
+    pub async fn buy_preimage_offer(
+        &self,
+        payment_hash: &bitcoin_hashes::sha256::Hash,
+        amount: &Amount,
+        mut rng: impl RngCore + CryptoRng,
+    ) -> Result<(minimint_api::TransactionId, ContractId)> {
+        let mut batch = DbBatch::new();
+
+        // Fetch offer for this payment hash
+        let offer: IncomingContractOffer = self.ln_client().get_offer(*payment_hash).await?;
+        if &offer.amount != amount || &offer.hash != payment_hash {
+            return Err(GatewayClientError::InvalidOffer);
+        }
+
+        // Inputs
+        let (mut coin_keys, coin_input) = self
+            .mint_client()
+            .create_coin_input(batch.transaction(), offer.amount)?;
+
+        // Outputs
+        let our_pub_key = secp256k1_zkp::schnorrsig::PublicKey::from_keypair(
+            &self.context.secp,
+            &self.context.config.redeem_key,
+        );
+        let contract = Contract::Incoming(IncomingContract {
+            hash: offer.hash,
+            encrypted_preimage: offer.encrypted_preimage.clone(),
+            decrypted_preimage: DecryptedPreimage::Pending,
+            gateway_key: our_pub_key,
+        });
+        let incoming_output =
+            minimint::transaction::Output::LN(ContractOrOfferOutput::Contract(ContractOutput {
+                amount: *amount,
+                contract: contract.clone(),
+            }));
+
+        // Submit transaction
+        let mut builder = TransactionBuilder::default();
+        builder.input(&mut coin_keys, Input::Mint(coin_input));
+        builder.output(incoming_output);
+        let tx = builder.build(&self.context.secp, &mut rng);
+        let mint_tx_id = self.context.api.submit_transaction(tx).await?;
+
+        self.context.db.apply_batch(batch).expect("DB error");
+
+        Ok((mint_tx_id, contract.contract_id()))
+    }
+
+    /// Claw back funds after outgoing contract that had invalid preimage
+    pub async fn claim_incoming_contract(
```

Comment:
It is a little confusing this has the same name but different behavior in the user and gateway clients.

Reply:
[At this point there is a contract inside the federation that has 3 possible states](https://github.com/fedimint/minimint/blob/162385f3e8e44c3f2a33a859fd5c63e49547d2fe/modules/minimint-ln/src/lib.rs#L234):

1. Preimage not decrypted yet
2. Spendable by the user if they provided a valid preimge
3. Spendable by the gateway if the preimage was invalid

The two `claim_incoming_contract` implementations do the same thing -- claim the contract. But the code looks slightly different because the `GatewayClient` lacks some of the helper methods added by your coin selection PR such as `submit_with_change`.

Reply:
I'm going to mark this resolved.